src/layer_env.rs

```diff
@@ -0,0 +1,882 @@
+/// ```
+#[derive(Eq, PartialEq, Debug)]
+pub struct LayerEnv {
+    all: LayerEnvDelta,
+    build: LayerEnvDelta,
+    launch: LayerEnvDelta,
+    process: HashMap<String, LayerEnvDelta>,
+
+    // Entries for the standard layer paths as described in the CNB spec:
+    // https://github.com/buildpacks/spec/blob/a9f64de9c78022aa7a5091077a765f932d7afe42/buildpack.md#layer-paths
+    // These cannot be set by the user itself and are only populated when a `LayerEnv` is read from
+    // disk by this library.
+    layer_paths_build: LayerEnvDelta,
+    layer_paths_launch: LayerEnvDelta,
+}
+
+impl LayerEnv {
+    /// Creates an empty LayerEnv that does not modify any environment variables.
+    ///
+    /// Entries can be added with the [LayerEnv::insert] function.
+    ///
+    /// # Example:
+    /// ```
+    /// use libcnb::layer_env::{LayerEnv, TargetLifecycle};
+    /// use libcnb::Env;
+    ///
+    /// let layer_env = LayerEnv::new();
+    /// let mut env = Env::new();
+    ///
+    /// let modified_env = layer_env.apply(TargetLifecycle::Build, &env);
+    /// assert_eq!(env, modified_env);
+    /// ```
+    pub fn new() -> Self {
+        LayerEnv {
+            all: LayerEnvDelta::new(),
+            build: LayerEnvDelta::new(),
+            launch: LayerEnvDelta::new(),
+            process: HashMap::new(),
+            layer_paths_build: LayerEnvDelta::new(),
+            layer_paths_launch: LayerEnvDelta::new(),
+        }
+    }
+
+    /// Applies this [`LayerEnv`] to the given [`Env`] for the given [target lifecycle](TargetLifecycle).
+    ///
+    /// # Example:
+    ///```
+    /// use libcnb::layer_env::{LayerEnv, TargetLifecycle, ModificationBehavior};
+    /// use libcnb::Env;
+    ///
+    /// let mut layer_env = LayerEnv::new();
+    /// layer_env.insert(TargetLifecycle::All, ModificationBehavior::Append, "VAR", "bar");
+    /// layer_env.insert(TargetLifecycle::All, ModificationBehavior::Default, "VAR2", "default");
+    ///
+    /// let mut env = Env::new();
+    /// env.insert("VAR", "foo");
+    /// env.insert("VAR2", "previous-value");
+    ///
+    /// let modified_env = layer_env.apply(TargetLifecycle::Build, &env);
+    /// assert_eq!(modified_env.get("VAR").unwrap(), "foobar");
+    /// assert_eq!(modified_env.get("VAR2").unwrap(), "previous-value");
+    /// ```
+    pub fn apply(&self, target: TargetLifecycle, env: &Env) -> Env {
+        let deltas = match target {
+            TargetLifecycle::All => vec![&self.all],
+            TargetLifecycle::Build => vec![&self.all, &self.build, &self.layer_paths_build],
+            TargetLifecycle::Launch => vec![&self.all, &self.launch, &self.layer_paths_launch],
+            TargetLifecycle::Process(process) => {
+                let mut process_deltas = vec![&self.all];
+                if let Some(process_specific_delta) = self.process.get(&process) {
+                    process_deltas.push(process_specific_delta);
+                }
```

Comment:
> It looks like if that happens then the process would get ALL of the env vars

I can see why you think that, but that's not the case. `self.all` contains all env var deltas that apply to all target lifecycles, not all deltas across all lifecycles.

With that out of the way, I don't think panicking is a good idea here. There is nothing wrong to apply the deltas for a process that doesn't have specific environment variables. If the user of this library passes in the wrong process type string, this is nothing the library should attempt to fix.

Reply:
I guess I'm struggling to see the use-case where someone would be using process-specific `.apply()` during the build, to understand what the user expectations might be? (I can see why we need to write out process-specific env files, but in my mind they mostly would be useful for runtime.)

Reply:
If there is a use-case during build it's most likely an edge-case. This PR includes support for applying them because it is mostly free in terms of required code. If we'd remove the ability to apply process specific environment deltas, we'd need more code as `TargetLifecycle` would be split up into two: one for specifying, one for applying. I can imagine the code being harder to follow too.